server/src/api/requests/quantity.ts

```diff
@@ -0,0 +1,119 @@
+import {DB} from "../../database";
+import {Item} from "../items/item";
+
+interface QuantitiesInStatus {
+    [statusName: string]: number;
+}
+
+interface ItemQuantities {
+    [itemId: string]: QuantitiesInStatus;
+}
+
+interface ItemQtyAvailable {
+    [itemId: string]: number;
+}
+
+export class Quantity {
+    public static async inStock(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+    }
+
+    public static async unreserved(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async availableForApproval(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        return this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+    }
+
+    public static async all(itemId: number = -1) {
+        const quantities: ItemQuantities = await Quantity.getQuantities(["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"], itemId);
+        const totalAvailable: ItemQtyAvailable = await Item.getTotalAvailable(itemId);
+
+        const qtyInStock: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["FULFILLED", "LOST", "DAMAGED"], itemId);
+        const qtyUnreserved: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+        const qtyAvailableForApproval: ItemQtyAvailable = this.getTotalAvailableLessStatuses(quantities, totalAvailable, ["APPROVED", "READY_FOR_PICKUP", "FULFILLED", "LOST", "DAMAGED"]);
+
+        return {
+            qtyInStock,
+            qtyUnreserved,
+            qtyAvailableForApproval
+        };
+    }
+
+    private static getTotalAvailableLessStatuses(quantities: ItemQuantities, totalAvailable: ItemQtyAvailable, statuses: RequestStatus[] = [], itemId: number = -1): ItemQtyAvailable {
+        const result = {};
+
+        for (const id in totalAvailable) {
+            if (totalAvailable.hasOwnProperty(id)) {
```

Comment:
probably throw an error if totalAvailable doesn't have the id right?

Reply:
I mean sure, but this hasOwnProperty is mostly a formality you need when using for-in